### PR TITLE
Update megatron fused softmax follow megatron-lm

### PR DESCRIPTION
**Pull request by yaoyu-33**
We updated megatron fused softmax in the following aspects:
1. We updated the limit of sequence length from 2048 to 4096 according to https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/fused_softmax.py#L171
2. We also enabled `mask=None` support in scaled_masked_softmax according to https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/fused_softmax.py#L84

### CODE REVIEW

File: csrc/megatron/scaled_softmax.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/* coding=utf-8
+ * Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
```
Comment (crcrpar) [inline csrc/megatron/scaled_softmax.cpp:2]:
```suggestion
 * Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
```

**Review by crcrpar — COMMENTED** [review]
how is the custom C++/CUDA extension of `scaled_softmax_cuda` built?
usually `setup.py` needs to be updated.
also, could you update https://github.com/NVIDIA/apex/blob/master/tests/L0/run_transformer/test_fused_softmax.py as well?

**Comment by yaoyu-33** [comment]
@crcrpar comments addressed

File: tests/L0/run_transformer/test_fused_softmax.py [hunk 2]
```diff
@@ -139,6 +139,84 @@ def test_autocast_fused_scale_mask_softmax(self):
...
+
+            expected = fused_fn(attention_scores_0, mask)
+            actual = torch_fn(attention_scores_1, mask)
+            self.assertEqual(actual, expected, msg=msg)
+
+            g0 = torch.rand_like(actual)
+            with torch.no_grad():
+                g1 = g0.clone()
+            expected.backward(g0)
+            actual.backward(g1)
```
Comment (crcrpar) [inline tests/L0/run_transformer/test_fused_softmax.py:188]: feels like these lines are based off of some existing cases but how about either checking grads or remove these lines?

Comment (yaoyu-33) [reply tests/L0/run_transformer/test_fused_softmax.py:188]: added grad check

Comment (yaoyu-33) [reply tests/L0/run_transformer/test_fused_softmax.py:188]: I think the purpose in other tests was just to test backward can work correctly

File: tests/L0/run_transformer/test_fused_softmax.py [hunk 3]
```diff
@@ -139,6 +139,86 @@ def test_autocast_fused_scale_mask_softmax(self):
...
+                actual = fused_fn(attention_scores_0, mask)
+                self.assertEqual(actual.dtype, dtype, msg=msg)
+            self.assertEqual(actual, expected, msg=msg)
+
+            g0 = torch.rand_like(actual)
+            with torch.no_grad():
+                g1 = g0.clone()
+            expected.backward(g0)
+            actual.backward(g1)
+            self.assertEqual(attention_scores_0.grad, attention_scores_1.grad, msg=msg)
```
Comment (crcrpar) [inline tests/L0/run_transformer/test_fused_softmax.py:220]: I'm seeing a dtype mismatch here

File: tests/L0/run_transformer/test_fused_softmax.py [hunk 4]
```diff
@@ -139,6 +139,86 @@ def test_autocast_fused_scale_mask_softmax(self):
...
+            expected = fused_fn(attention_scores_0, mask)
+            actual = torch_fn(attention_scores_1, mask)
+            self.assertEqual(actual, expected, msg=msg)
+
+            g0 = torch.rand_like(actual)
+            with torch.no_grad():
+                g1 = g0.clone()
+            expected.backward(g0)
+            actual.backward(g1)
+            self.assertEqual(attention_scores_0.grad, attention_scores_1.grad, msg=msg)
```
Comment (crcrpar) [inline tests/L0/run_transformer/test_fused_softmax.py:189]: and this check complains that 1% elements have numerical mismatch

**Review by crcrpar — COMMENTED** [review]
generally okay but could you maybe tweak tests a bit?

File: tests/L0/run_transformer/test_fused_softmax.py (hunk 4 above)
Comment (yaoyu-33) [reply tests/L0/run_transformer/test_fused_softmax.py:189]: lol, do we know why? I am not familiar with this
